### ISSUE TITLE
How to choose number of epochs in Conformer-Transducer (RNNT) pretrained model finetuning

### ISSUE THREAD

**Issue by manjuke** [NVIDIA/NeMo#5806]
Hi All,

I have taken conformer-transducer (RNNT) & conformer CTC pretrained English medium models, and finetuned them for Tamil dataset. Finetuned CTC has decent WERs, whereas RNNT models have unacceptable WERs. Both were trained for 18 epochs. The CTC finetuning has consistently exponentially decreasing Validation loss. However, validation loss of RNNT consistently increased after 5th epoch. PFA RNNT val loss snapshot.

![image](https://user-images.githubusercontent.com/6142443/212663699-62081f70-cf12-418d-9e4d-52b1464aa57b.png)

So, in this case, 
Question 1: Do you suggest me to train only for 5 epochs while using RNNT and use the saved model for decoding?
Question 2: Is there any way like any parameter to be set either in exp_manager or trainer, so that it would automatically save the check point with best WER
Question 3: Am I doing anything wrong here. Should I need to check/change anything to make RNNT validation error to decrease.

Note: We have done several experiments with Conformer-CTC and they seem to work without any issues with same setup (like no. of epochs etc). However, we are exploring the RNNT for the first time. So, not sure if something is missing.
Please suggest @titu1994  Thanks.

**Comment by VahidooX** [comment]
>>Do you suggest me to train only for 5 epochs while using RNNT and use the saved model for decoding?
The RNNT variant is more likely to overfit your data specially for small dataset comparing to the CTC model. You may try the followings:
1-get more data
2-increase weight_decay to 1e-2 or even 1e-1
3-limit the number of epochs
4-how many time maskings you have? 10? at least use 10x time maskings.
5-You may try noise-robust training specially if you want to cover such cases
6-Isn't this model a better starting checkpoint comparing to English one? 
https://catalog.ngc.nvidia.com/orgs/nvidia/teams/nemo/models/stt_hi_conformer_ctc_medium

>>Is there any way like any parameter to be set either in exp_manager or trainer, so that it would automatically save the check point with best WER
In the default config files, the checkpoint_callback_params section is already set to save the top 5 best checkpoint based on val_wer. You may change it to save it best on other metrics. You should see the 5 best ones already saved under the checkpoint folder. Then you may use checkpoint averaging script under the scripts folder to create an averaged checkpoint from those the best.

**Comment by manjuke** [comment]
Hi @VahidooX 

Thanks for the response. I will explore points 1 to 5 and try to retrain the model.

My response to Point 6: I have already explored "stt_hi_conformer_ctc_medium" while building ctc based models for Hindi. But, since I wanted to explore RNNT models currently I am using "stt_en_conformer_transducer_medium"  pretrained models. And, also our data has both Tamil & English characters, Hence I felt it would appropriate to use "English" models over "Hindi". Please let me know, if I am missing anything here.

Regarding the checkpoint: Ya true that "save_top_k: 5" is set. However, in the checkpoints folder, I am seeing only the checkpoint corresponds to last epoch & last step (and it is not even the best checkpoint as per the validation loss plots). Pls suggest.

This is how my trainer looks like:
trainer = pl.Trainer(strategy=DDPStrategy(find_unused_parameters=False), 
                     accelerator='gpu',  
                     #devices=[1,2,3],
                     devices=-1,
                     auto_lr_find = True, 
                     auto_scale_batch_size = True, 
                     val_check_interval = 0.5, 	
                     default_root_dir = rootDir,                  
                     max_epochs=18
                     )

Thanks.

**Comment by manjuke** [comment]
For Point 6, Just wanted to know is it possible to use conformer CTC pretrained models to train RNNT models? If so, could you please point me to any resources/notebooks for the same. Thanks @VahidooX

**Comment by titu1994** [comment]
Yes it's possible. We don't do it normally cause our conformer CTC encoder is not 1:1 for RNNT and so you can't load the weights directly. But in theory yes if you have same model encoder for CTC and RNNT then that can be done.

**Comment by manjuke** [comment]
Thanks @titu1994
